### PR TITLE
[sitecore-jss-react][sitecore-jss-nextjs] Prevent hydration error with dynamic components

### PR DESCRIPTION
**Pull request by art-alexeyenko**
- [x] This PR follows the [Contribution Guide](https://github.com/Sitecore/jss/blob/dev/CONTRIBUTING.md)
- [x] Changelog updated - TBA
- [ ] Upgrade guide entry added

## Description / Motivation
Fixes hydration issues when ErrorBoundary's Suspense wraps nextjs's dynamic() wrapper.
dynamic() components will have a `render` prop. If it's present - the component will not be wrapped in ErrorBoundary's `Suspense` preventing intermittent hydration problems.

## Testing Details
- [x] Unit Test Added 
- [x] Manual Test/Other (Please elaborate) - Errors are successfully captured in connected mode

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/sitecore-jss-react/src/components/PlaceholderCommon.tsx [hunk 1]
```diff
@@ -227,6 +227,7 @@ export class PlaceholderCommon<T extends PlaceholderProps> extends React.Compone
           component = () => <></>;
         } else {
           component = this.getComponentForRendering(componentRendering);
+          console.log(component);
```
Comment (sc-addypathania) [inline packages/sitecore-jss-react/src/components/PlaceholderCommon.tsx:230]: Please remove the log statement.

File: packages/sitecore-jss-react/src/components/PlaceholderCommon.tsx [hunk 2]
```diff
@@ -267,11 +268,27 @@ export class PlaceholderCommon<T extends PlaceholderProps> extends React.Compone
...
+              componentLoadingMessage={this.props.componentLoadingMessage}
+              type={type}
+              isDynamic={(component as JssComponentType).isDynamic}
+              {...rendered.props}
+            >
+              {rendered}
+            </ErrorBoundary>
+          );
+        }
+
```
Comment (sc-addypathania) [inline packages/sitecore-jss-react/src/components/PlaceholderCommon.tsx:291]: with this refactor does this mean `PlaceholderMetadata` component would not be included in the `ErrorBoundary`?

File: packages/sitecore-jss-react/src/components/ErrorBoundary.tsx [hunk 3]
```diff
@@ -78,6 +79,11 @@ class ErrorBoundary extends React.Component<ErrorBoundaryProps> {
       }
     }
 
+    // do not apply suspense on already dynamic components
+    if (this.props.isDynamic) {
```
Comment (sc-addypathania) [inline packages/sitecore-jss-react/src/components/ErrorBoundary.tsx:83]: Please add a test case.

File: packages/sitecore-jss-react/src/components/PlaceholderCommon.tsx (hunk 2 above)
Comment (art-alexeyenko) [reply packages/sitecore-jss-react/src/components/PlaceholderCommon.tsx]: Yes. The main goal of ErrorBoundary is to alert to problems with customer's and app's components. With this change the focus is moved on these components specifically and our built-in stuff will not affect error boundaries.

Comment (sc-addypathania) [reply packages/sitecore-jss-react/src/components/PlaceholderCommon.tsx]: but what if an error occurs in the `PlaceholderMetadata` component?

Comment (art-alexeyenko) [reply packages/sitecore-jss-react/src/components/PlaceholderCommon.tsx]: Then we have a larger problem on our side and each page of the site stops working in editing mode.
Which would be roughly the same behavior if PlaceholderMetadata errors are handled by ErrorBoundary - the content will not be shown.

File: packages/sitecore-jss-react/src/components/ErrorBoundary.test.tsx [hunk 4]
```diff
@@ -243,6 +243,50 @@ describe('ErrorBoundary', () => {
...
+          <ItsADynamicComponent />
+        </ErrorBoundary>
+      );
+      expect(rendered.text()).to.equal(loading);
+    });
+
+    it('should not render Suspense and default loading message when wrapping a dynamic component', async () => {
+      // mount fails with lazy component and no suspense
+      const rendered = mount(
+        <Suspense>
```
Comment (sc-addypathania) [inline packages/sitecore-jss-react/src/components/ErrorBoundary.test.tsx:278]: Does this mean we have to wrap all the dynamic components with <Suspense>?

Comment (art-alexeyenko) [reply packages/sitecore-jss-react/src/components/ErrorBoundary.test.tsx:278]: for tests, yes
mount here refers to enzyme mount() specifically

**Review by sc-addypathania — APPROVED** [review]
👍🏼

File: packages/sitecore-jss-react/src/components/PlaceholderCommon.tsx [hunk 5]
```diff
@@ -267,11 +267,27 @@ export class PlaceholderCommon<T extends PlaceholderProps> extends React.Compone
...
         };
 
-        const rendered = React.createElement<{ [attr: string]: unknown }>(
+        let rendered = React.createElement<{ [attr: string]: unknown }>(
           component as React.ComponentType,
           this.props.modifyComponentProps ? this.props.modifyComponentProps(finalProps) : finalProps
         );
 
+        const type = rendered.props.type === 'text/sitecore' ? rendered.props.type : '';
+        if (!isEmpty) {
```
Comment (illiakovalenko) [inline packages/sitecore-jss-react/src/components/PlaceholderCommon.tsx:276]: @art-alexeyenko Previously we had some comments here explaining problems, it should be reverted back

File: packages/sitecore-jss-react/src/components/PlaceholderCommon.tsx [hunk 6]
```diff
@@ -267,11 +267,27 @@ export class PlaceholderCommon<T extends PlaceholderProps> extends React.Compone
           rendering: componentRendering,
         };
 
-        const rendered = React.createElement<{ [attr: string]: unknown }>(
+        let rendered = React.createElement<{ [attr: string]: unknown }>(
           component as React.ComponentType,
           this.props.modifyComponentProps ? this.props.modifyComponentProps(finalProps) : finalProps
         );
 
+        const type = rendered.props.type === 'text/sitecore' ? rendered.props.type : '';
```
Comment (illiakovalenko) [inline packages/sitecore-jss-react/src/components/PlaceholderCommon.tsx:275]: @art-alexeyenko Should be added under _if_ statement

File: packages/sitecore-jss-react/src/components/PlaceholderCommon.tsx (hunk 5 above)
Comment (illiakovalenko) [inline packages/sitecore-jss-react/src/components/PlaceholderCommon.tsx:276]: @art-alexeyenko Add a comment here explaining that components are not wrapped in _Chromes Edit Mode_

**Pull request merged**